util/EIP-2124/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+	where
+		PF: IntoIterator<Item = BlockNumber>,
+		NF: IntoIterator<Item = BlockNumber>,
+	{
+		let genesis_fork_hash = ForkHash::from(genesis);
+		Self {
+			head,
+			past_forks: past_forks.into_iter().fold((btreemap! { 0 => genesis_fork_hash }, genesis_fork_hash), |(mut acc, base_hash), block| {
+				let fork_hash = base_hash + block;
+				acc.insert(block, fork_hash);
+				(acc, fork_hash)
+			}).0,
+			next_forks: next_forks.into_iter().collect(),
+		}
+	}
+
+	fn current_fork_hash(&self) -> ForkHash {
+		*self.past_forks.values().next_back().unwrap()
+	}
+
+	fn future_fork_hashes(&self) -> Vec<ForkHash> {
+		self.next_forks.iter().fold((Vec::new(), self.current_fork_hash()), |(mut acc, hash), fork| {
+			let next = hash + *fork;
+			acc.push(next);
+			(acc, next)
+		}).0
+	}
+
+	pub fn insert_past_fork(&mut self, height: BlockNumber) {
+		self.past_forks.insert(height, self.current_fork_hash() + height);
+	}
+
+	pub fn insert_next_fork(&mut self, height: BlockNumber) {
+		self.next_forks.insert(height);
+	}
+
+	pub fn promote_next_fork(&mut self, height: BlockNumber) -> bool {
+		let promoted = self.next_forks.remove(&height);
+		if promoted {
+			self.insert_past_fork(height);
+		}
+		promoted
+	}
+
+	pub fn is_valid(&self, fork_id: ForkId) -> Result<(), RejectReason> {
+		// 1) If local and remote FORK_HASH matches...
+		if self.current_fork_hash() == fork_id.hash {
+			if fork_id.next == 0 {
+				// 1b
+				return Ok(())
+			}
+
+			//... compare local head to FORK_NEXT.
+			if self.head < fork_id.next {
+				return Ok(())
+			} else {
+				return Err(RejectReason::LocalIncompatibleOrStale)
+			}
+		}
+		
+		// 2) If the remote FORK_HASH is a subset of the local past forks...
+		let mut it = self.past_forks.iter();
+		while let Some((_, hash)) = it.next() {
+			if *hash == fork_id.hash {
+				// ...and the remote FORK_NEXT matches with the locally following fork block number, connect.
+				if let Some((actual_fork_block, _)) = it.next() {
+					if *actual_fork_block == fork_id.next {
+						return Ok(())
+					} else {
+						return Err(RejectReason::RemoteStale);
+					}
+				}
+
+				break;
+			}
+		}
+
+		// 3) If the remote FORK_HASH is a superset of the local past forks and can be completed with locally known future forks, connect.
+		for future_fork_hash in self.future_fork_hashes() {
+			if future_fork_hash == fork_id.hash {
+				return Ok(())
+			}
+		}
+
+		// 4) Reject in all other cases
+		Err(RejectReason::LocalIncompatibleOrStale)
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use hex_literal::hex;
+
+	const GENESIS_HASH: &str = "d4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3";
+	const BYZANTIUM_FORK_HEIGHT: BlockNumber = 4370000;
+	const PETERSBURG_FORK_HEIGHT: BlockNumber = 7280000;
+
+	#[test]
+	fn test_forkhash() {
+		let mut fork_hash = ForkHash::from(GENESIS_HASH.parse::<H256>().unwrap());
+		assert_eq!(fork_hash.0, 0xfc64ec04);
```

Comment:
Are these values from the EIP test vectors? If so it would be good with a comment to that effect.

Reply:
done